aesd-char-driver/main.c

```diff
@@ -228,12 +229,147 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
 
 
 
+/**
+ * @brief Custom llseek implementation to support seek operations
+ *
+ * @param filp     File pointer
+ * @param off      Offset value
+ * @param whence   Positional type (SEEK_SET, SEEK_CUR, SEEK_END)
+ * @return New file position on success, or appropriate error code
+ */
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+    struct aesd_dev *dev = filp->private_data;
+    loff_t new_pos = 0;
+    size_t total_size = 0;
+    struct aesd_buffer_entry *entry;
+    uint8_t index;
+
+    if (mutex_lock_interruptible(&dev->mutex_lock))
+    {
+    	printk("mutex lock unsuccess\n");
+        return -ERESTARTSYS;
+    }
+
+    /*switch (whence)
+    {
+    case SEEK_SET:
+    	printk("seek set, %u\n", new_pos);
+        new_pos = off;
+        break;
+    case SEEK_CUR:
+    	printk("seek cur %u\n", new_pos);
+        new_pos = filp->f_pos + off;
+        break;
+    case SEEK_END:
+        // Calculate the total size of all entries in the circular buffer*/
+        //printk("seek end %u\n", new_pos);
+        AESD_CIRCULAR_BUFFER_FOREACH(entry, &dev->buf, index)
+        {
+            total_size += entry->size;
+        }
+        //new_pos = total_size + off;
+        new_pos = fixed_size_llseek(filp, off, whence, total_size);
+        //break;
+    /*default:
+    	printk("default case\n");
+        new_pos = -EINVAL;
+    }*/
+
+    if (new_pos < 0 || new_pos > total_size)
+    {
+        // Seek position is out of bounds
+        printk("out of range pos %u\n", new_pos);
+        new_pos = -EINVAL;
+    }
+    else
+    {
+        // Update the file position
+        filp->f_pos = new_pos;
+        printk("updatted pos %u\n", new_pos);
+    }
+
+    mutex_unlock(&dev->mutex_lock);
+    return new_pos;
+}
+
+/**
+ * @brief ioctl command handler for AESDCHAR_IOCSEEKTO
+ *
+ * @param filp     File pointer
+ * @param cmd      ioctl command (AESDCHAR_IOCSEEKTO)
+ * @param arg      User-provided buffer containing command index and offset
+ * @return 0 on success, or appropriate error code
+ */
+long aesd_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+    struct aesd_dev *dev = filp->private_data;
+    int retval = 0;
+    struct aesd_seekto seekto;
+    uint8_t i;
+    size_t new_pos = 0;
+    uint8_t index = dev->buf.out_offs;
+    struct aesd_buffer_entry *entry;
+
+    if (_IOC_TYPE(cmd) != AESD_IOC_MAGIC)
+        return -ENOTTY;
+
+    if (_IOC_NR(cmd) != AESDCHAR_IOCSEEKTO)
+        return -ENOIOCTLCMD;
+
+    if (mutex_lock_interruptible(&dev->mutex_lock))
```

Comment:
@akka2103 you could shorten this locking interval, you only need to lock accesses to dev.